go/extractor/project/project.go

```diff
@@ -42,109 +124,366 @@ func checkDirsNested(inputDirs []string) (string, bool) {
+	}
+
+	// If there are no `go.mod` files at all, create one in line with https://go.dev/blog/migrating-to-go-modules
+	if totalModuleFiles == 0 {
+		// Check for other, legacy package managers
+		if util.FileExists("Gopkg.toml") {
+			if emitDiagnostics {
+				diagnostics.EmitGopkgTomlFound()
+			}
+			log.Println("Found Gopkg.toml, using dep instead of go get")
+			goWorkspaces = []GoWorkspace{{
+				BaseDir: ".",
+				DepMode: Dep,
+				ModMode: ModUnset,
+			}}
+			totalModuleFiles = 0
+			return
 		}
+
+		if util.FileExists("glide.yaml") {
+			if emitDiagnostics {
+				diagnostics.EmitGlideYamlFound()
+			}
+			log.Println("Found glide.yaml, using Glide instead of go get")
+			goWorkspaces = []GoWorkspace{{
+				BaseDir: ".",
+				DepMode: Glide,
+				ModMode: ModUnset,
+			}}
+			totalModuleFiles = 0
+			return
+		}
+
+		// If we have no `go.mod` files, then the project appears to be a legacy project without
+		// a `go.mod` file. Check that there are actually Go source files before initializing a module
+		// so that we correctly fail the extraction later.
+		if !util.FindGoFiles(".") {
+			goWorkspaces = []GoWorkspace{{
+				BaseDir: ".",
+				DepMode: GoGetNoModules,
+				ModMode: ModUnset,
+			}}
+			totalModuleFiles = 0
+			return
+		}
+
+		goWorkspaces = []GoWorkspace{{
+			BaseDir: ".",
+			DepMode: GoGetNoModules,
+			ModMode: getModMode(GoGetWithModules, "."),
+		}}
+		totalModuleFiles = 0
+		return
 	}
 
-	baseDir, useGoMod := findGoModFiles(emitDiagnostics)
-	if useGoMod {
-		log.Println("Found go.mod, enabling go modules")
-		return GoGetWithModules, baseDir
+	// Get the paths to all `go.mod` files
+	i := 0
+	goModPaths := make([]string, totalModuleFiles)
+
+	for _, goWorkspace := range goWorkspaces {
+		for _, goModule := range goWorkspace.Modules {
+			goModPaths[i] = goModule.Path
+			i++
+		}
 	}
 
-	if util.FileExists("Gopkg.toml") {
+	goModDirs := util.GetParentDirs(goModPaths)
+	straySourceFiles := util.GoFilesOutsideDirs(".", goModDirs...)
+	if len(straySourceFiles) > 0 {
 		if emitDiagnostics {
-			diagnostics.EmitGopkgTomlFound()
+			diagnostics.EmitGoFilesOutsideGoModules(goModPaths)
+		}
+
+		// We need to initialise Go modules for the stray source files. Our goal is to initialise
+		// as few Go modules as possible, in locations which do not overlap with existing Go
+		// modules.
+		for _, straySourceFile := range straySourceFiles {
+			path := "."
+			components := strings.Split(filepath.Dir(straySourceFile), string(os.PathSeparator))
+
+			for _, component := range components {
+				path = filepath.Join(path, component)
+
+				// Try to initialize a `go.mod` file automatically for the stray source files.
+				if !slices.Contains(goModDirs, path) {
+					goWorkspaces = append(goWorkspaces, GoWorkspace{
+						BaseDir: path,
+						DepMode: GoGetNoModules,
+						ModMode: ModUnset,
+					})
+					goModDirs = append(goModDirs, path)
+					break
+				}
+			}
+		}
+
+		return
```

Comment:
This preserves the old behaviour, so it should stay like this. But I wonder if in future we might want to remove this `return` so that we get some other diagnostics in this case as well.